articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,227 @@
+   - An x86 computer running Xenomai or RT_PREEMPT that receives the sensor data over Ethernet or USB and processes the output in soft or firm real-time, but also runs separate non-RT processes for user interaction.
+
+2. Implement the system with ROS 2.
+
+3. Profile the performance of the system.
+
+4. Identify solutions to improve the performance of the system, e.g. by exposing more concurrency primitives through the appropriate abstraction.
+
+5. Adapt the ROS 2 API or underlying libraries to implement these solutions and to improve the experience of writing real-time code.
+
+6. Refactor the system for the new API if necessary and repeat.
+
+## Software architecture challenges and solutions
+
+Wherever ROS 2 libraries expose sources of nondeterminism, there should be a path that exposes an alternative.
+This section evaluates the current state of ROS 2 and rclcpp, predicts challenges to fulfilling real-time requirements, and presents possible solutions.
+
+### Memory Allocation
+
+Poor memory management can create performance issues due to page faults and nondeterministic algorithms for dynamically allocating memory.
+Although C++ has an advantage over many programming languages because memory management is exposed, the default memory allocator for operator `new` and dynamically sized data structures is not well-suited to real-time applications (see [Composing High-Performance Memory Allocators](https://people.cs.umass.edu/~emery/pubs/berger-pldi2001.pdf)).
+Dynamically-sized data structures in the C++ standard library offer an allocator interface, such as `std::vector` (see [cppreference.com](http://en.cppreference.com/w/cpp/concept/Allocator)).
+C++11 greatly improved this interface by introducing [`std::allocator_traits`](http://en.cppreference.com/w/cpp/memory/allocator_traits), which constructs an allocator compatible with the standard allocator interface from a minimal set of requirements
+To allow for optimal performance in `rclcpp`, the client library should expose custom allocators for publishers, subscribers, etc. that are propagated to standard library structures used by these entities.
+
+As of the writing of this document (1/6/16), selecting custom allocators in `rclcpp` is almost fully implemented.
+An example of the ROS 2 allocator API using the Two Level Segregate Fit allocator is available [on Github](https://github.com/ros2/realtime_support/blob/master/tlsf_cpp/example/allocator_example.cpp).
+
+The following work remains to be done:
+
+- Improve the interface in rclcpp such that the user need only provide an "allocate" function, a "deallocate" function, and an optional pointer to the allocator state (this interface is currently outlined in the rcl API).
+- Custom allocators for services/clients.
+- Custom allocators for parameters.
+- Fully implement and test the allocator interface in rcl, rclc, and rmw.
+
+### Scheduling and synchronization
```

Comment:
You're accumulating a lot of knowledge about implementing real-time, so it would be good to preserve it in a ROS-specific fashion so that it is immediately accessible to ROS users. At an appropriate point in the future, it would be worth adding some documentation to the wiki giving an example of correct use of atomics and real-time-safe synchronisation primitives. Some people may find this useful for writing their nodes, but the use case I'm thinking of is people writing custom executors that meet their (assumed to be) unique needs.